### PR TITLE
Migrate sqlite data to etcd when initializing the cluster

### PR DESCRIPTION
**Pull request by ibuildthecloud**
* https://github.com/k3s-io/k3s/issues/3982

### CODE REVIEW

File: pkg/etcd/etcd.go [hunk 1]
```diff
@@ -495,11 +501,68 @@ func GetAdvertiseAddress(advertiseIP string) (string, error) {
...
+	} else if err != nil {
+		return err
+	}
+
+	logrus.Infof("Migrating content from sqlite to etcd")
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	_, err = endpoint2.Listen(ctx, endpoint2.Config{
```
Comment (dweomer) [inline pkg/etcd/etcd.go:531]: Are we leaking this listener on the kine socket when we start a new cluster with sqlite data available?

Comment (ibuildthecloud) [reply pkg/etcd/etcd.go:531]: That's the reason for the `defer cancel()` above.  The only way to close the listener is to close the context.

**Comment by ibuildthecloud** [comment]
Important note.  If migrations fails the startup will fail, but on subsequent start migration will be skipped because the etcd datastore is already initialized.  If one wants to retry/restart the migration they would have to delete the server/db/etcd folder.

File: pkg/etcd/etcd.go (hunk 1 above)
Comment (dweomer) [reply pkg/etcd/etcd.go:531]: Ah, pardon the low-caffeine "review"

**Review by dweomer — APPROVED** [review]

**Comment by brandond** [comment]
Can we move the sqlite database file aside (into a .old/.backup/etc directory) when migration succeeds? Leaving it around seems like it might cause some odd things when users are trying to do a cluster reset or restore.

**Comment by ibuildthecloud** [comment]
Updated so that after migration the state.db file is moved to state.db.migrated

**Comment by cjellick** [comment]
Not knowing the code, it isn't obvious to me how a user opts into this. Is it gated by the `--cluster-init` flag?

**Review by galal-hussein — APPROVED** [review]

**Comment by galal-hussein** [comment]
> Not knowing the code, it isn't obvious to me how a user opts into this. Is it gated by the `--cluster-init` flag?

If I understand correctly it will only migrate if it finds the sqlite database, this happens indeed when user passes --cluster-init to start etcd

File: pkg/etcd/etcd.go [hunk 2]
```diff
@@ -495,11 +501,68 @@ func GetAdvertiseAddress(advertiseIP string) (string, error) {
...
+
+	for _, value := range values {
+		logrus.Infof("Migrating etcd key %s", value.Key)
+		_, err := etcdClient.Put(ctx, string(value.Key), string(value.Data))
+		if err != nil {
+			return err
+		}
+	}
+
+	return os.Rename(sqliteFile(e.config), sqliteFile(e.config)+".migrated")
```
Comment (brandond) [inline pkg/etcd/etcd.go:565]: 💯

File: pkg/etcd/etcd.go [hunk 3]
```diff
@@ -495,11 +501,68 @@ func GetAdvertiseAddress(advertiseIP string) (string, error) {
...
+	err := e.cluster(ctx, reset, executor.InitialOptions{
 		AdvertisePeerURL: fmt.Sprintf("https://%s:2380", e.address),
 		Cluster:          fmt.Sprintf("%s=https://%s:2380", e.name, e.address),
 		State:            "new",
 	})
+	if err != nil {
+		return err
+	}
+	if err := e.migrateFromSQLite(ctx); err != nil {
+		return fmt.Errorf("failed to migration content from sqlite to etcd: %w", err)
```
Comment (brandond) [inline pkg/etcd/etcd.go:513]:
```suggestion
		return fmt.Errorf("failed to migrate content from sqlite to etcd: %w", err)
```

**Review by brandond — CHANGES_REQUESTED** [review]

**Comment by brandond** [comment]
code lgtm, one nit on error message grammar

**Review by brandond — APPROVED** [review]

**Comment by xavier83** [comment]
@brandond bump. can we see this in the next release? waiting for HA control plane migration.. :smiley:

**Pull request merged**